monai/data/wsi_reader.py

```diff
@@ -623,34 +716,71 @@ def get_level_count(wsi) -> int:
         """
         return len(wsi.pages)
 
-    @staticmethod
-    def get_size(wsi, level: int) -> Tuple[int, int]:
+    def get_size(self, wsi, level: Optional[int] = None) -> Tuple[int, int]:
         """
         Returns the size (height, width) of the whole slide image at a given level.
 
         Args:
             wsi: a whole slide image object loaded from a file
-            level: the level number where the size is calculated
+            level: the level number where the size is calculated. If not provided the default level (from `self.level`)
+                will be used.
 
         """
+        if level is None:
+            level = self.level
+
         return (wsi.pages[level].imagelength, wsi.pages[level].imagewidth)
 
-    @staticmethod
-    def get_downsample_ratio(wsi, level: int) -> float:
+    def get_downsample_ratio(self, wsi, level: Optional[int] = None) -> float:
         """
         Returns the down-sampling ratio of the whole slide image at a given level.
 
         Args:
             wsi: a whole slide image object loaded from a file
-            level: the level number where the size is calculated
+            level: the level number where the size is calculated. If not provided the default level (from `self.level`)
+                will be used.
 
         """
+        if level is None:
+            level = self.level
+
         return float(wsi.pages[0].imagelength) / float(wsi.pages[level].imagelength)
 
-    def get_file_path(self, wsi) -> str:
+    @staticmethod
+    def get_file_path(wsi) -> str:
         """Return the file path for the WSI object"""
         return str(abspath(wsi.filehandle.path))
 
+    def get_mpp(self, wsi, level: Optional[int] = None) -> Tuple[float, float]:
+        """
+        Returns the micro-per-pixel resolution of the whole slide image at a given level.
+
+        Args:
+            wsi: a whole slide image object loaded from a file
+            level: the level number where the size is calculated. If not provided the default level (from `self.level`)
+                will be used.
+
+        """
+        if level is None:
+            level = self.level
+
+        unit = wsi.pages[level].tags["ResolutionUnit"].value
+        if unit == unit.CENTIMETER:
+            factor = 10000.0
+        elif unit == unit.MILLIMETER:
+            factor = 1000.0
+        elif unit == unit.MICROMETER:
+            factor = 1.0
+        elif unit == unit.INCH:
+            factor = 25400.0
+        else:
+            raise ValueError(f"The resolution unit is not a valid tiff resolution or missing: {unit.name}")
+
+        # Here x and y resolutions are rational numbers so each of them is represented by a tuple.
+        yres = wsi.pages[level].tags["YResolution"].value
+        xres = wsi.pages[level].tags["XResolution"].value
+        return (factor * yres[1] / yres[0], factor * xres[1] / xres[0])
```

Comment:
Hi @drbeh, the first number represents the numerator of a fraction and the second is the denominator. It seems like  you've got the numerator and denominator backwards.

Reply:
And did you missing to multiply by the `downsample_ratio` here? I have tried three backend on the same image and get three different micro-per-pixel, is there something wrong？

Reply:
Hi @KumoLiu, thanks for comments. The calculation of mpp is correct. `XResolution` and `YResolution` have inverse relation to mpp so the fraction looks fine. The reason you are getting different numbers have two reason: 1) the test image is not a fully valid TIFF file (that's why I only check at one level) and 2) there was a bug in cucim that it is resolved in the new version.
If you like you can check with one of the images here: https://drive.google.com/drive/folders/0BzsdkU4jWx9BWk11WEtZZUNFY0U?resourcekey=0-U0E7SyHPJeQd77VAi3z15Q